File: api/Services/SqsService.cs
using Amazon.Runtime.CredentialManagement;
using Amazon.SQS;
using Amazon.SQS.Model;
using DueDiligenceRiskTakerTool.Api.Models;

namespace DueDiligenceRiskTakerTool.Api.Services;

/// <summary>
/// SQS mode: list queues, send a crafted message, purge a queue.
/// Send/purge are MUTATING actions and may run against prod queues — the controller
/// enforces a typed-queue-name confirmation for purge. Never log message bodies or
/// attribute values (may contain Due Diligence domain data) — log queue names + ids only.
/// </summary>
public class SqsService
{
    private readonly IAmazonSQS _sqs;
    private readonly ILogger<SqsService> _logger;

    public SqsService(IAmazonSQS sqs, ILogger<SqsService> logger)
    {
        _sqs = sqs;
        _logger = logger;
    }

    public async Task<List<SqsQueue>> ListQueuesAsync(string? profile = null)
    {
        if (string.IsNullOrEmpty(profile))
            return await RunListAsync(_sqs);

        using var client = CreateClientForProfile(profile);
        return await RunListAsync(client);
    }

    private static async Task<List<SqsQueue>> RunListAsync(IAmazonSQS client)
    {
        var queues = new List<SqsQueue>();
        string? token = null;
        do
        {
            var resp = await client.ListQueuesAsync(new ListQueuesRequest { MaxResults = 1000, NextToken = token });
            foreach (var url in resp.QueueUrls)
            {
                var name = url.Split('/').Last();
                queues.Add(new SqsQueue { Url = url, Name = name, Fifo = name.EndsWith(".fifo") });
            }
            token = resp.NextToken;
        } while (!string.IsNullOrEmpty(token));

        return queues.OrderBy(q => q.Name).ToList();
    }

    public async Task<SqsSendResponse> SendMessageAsync(SqsSendRequest request)
    {
        using var client = string.IsNullOrEmpty(request.Profile) ? null : CreateClientForProfile(request.Profile);
        var sqs = client ?? _sqs;

        var sendRequest = new SendMessageRequest
        {
            QueueUrl = request.QueueUrl,
            MessageBody = request.Body,
        };

        if (request.MessageAttributes is { Count: > 0 })
        {
            sendRequest.MessageAttributes = request.MessageAttributes
                .Where(a => !string.IsNullOrWhiteSpace(a.Name))
                .ToDictionary(
                    a => a.Name,
                    a => new MessageAttributeValue
                    {
                        DataType = string.IsNullOrWhiteSpace(a.DataType) ? "String" : a.DataType,
                        StringValue = a.Value ?? string.Empty,
                    });
        }

        if (!string.IsNullOrWhiteSpace(request.MessageGroupId))
            sendRequest.MessageGroupId = request.MessageGroupId;
        if (!string.IsNullOrWhiteSpace(request.MessageDeduplicationId))
            sendRequest.MessageDeduplicationId = request.MessageDeduplicationId;

        var resp = await sqs.SendMessageAsync(sendRequest);
        _logger.LogInformation("SQS send to {Queue} → messageId {MessageId}",
            request.QueueUrl.Split('/').Last(), resp.MessageId);

        return new SqsSendResponse { MessageId = resp.MessageId, SequenceNumber = resp.SequenceNumber };
    }

    public async Task PurgeQueueAsync(string queueUrl, string? profile = null)
    {
        using var client = string.IsNullOrEmpty(profile) ? null : CreateClientForProfile(profile);
        var sqs = client ?? _sqs;

        await sqs.PurgeQueueAsync(new PurgeQueueRequest { QueueUrl = queueUrl });
        _logger.LogWarning("SQS PURGE issued for {Queue}", queueUrl.Split('/').Last());
    }

    // Short-lived client using a named AWS profile (region from ~/.aws/config; falls back to eu-west-1).
    private static AmazonSQSClient CreateClientForProfile(string profileName)
    {
        var chain = new CredentialProfileStoreChain();

        if (!chain.TryGetAWSCredentials(profileName, out var credentials))
            throw new InvalidOperationException(
                $"AWS profile '{profileName}' not found or has no usable credentials. " +
                "Check ~/.aws/credentials and ~/.aws/config.");

        chain.TryGetProfile(profileName, out var profile);
        var region = profile?.Region ?? Amazon.RegionEndpoint.EUWest1;

        return new AmazonSQSClient(credentials, region);
    }
}

File: applications/DueDiligence.RiskTakerTool.WebUI/services/sqsService.ts
const API_BASE = process.env.NEXT_PUBLIC_API_URL ?? 'http://localhost:5000';

export interface SqsQueue {
  url: string;
  name: string;
  fifo: boolean;
}

export interface SqsMessageAttribute {
  name: string;
  dataType: 'String' | 'Number';
  value: string;
}

export interface SqsSendRequest {
  queueUrl: string;
  body: string;
  messageAttributes?: SqsMessageAttribute[];
  messageGroupId?: string | null;
  messageDeduplicationId?: string | null;
  profile?: string | null;
}

export interface SqsSendResponse {
  messageId: string;
  sequenceNumber?: string | null;
}

async function post<T>(path: string, body: unknown): Promise<T> {
  const res = await fetch(`${API_BASE}${path}`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  });
  if (!res.ok) {
    const text = await res.text();
    throw new Error(`API ${res.status}: ${text}`);
  }
  return res.json();
}

export async function fetchQueues(profile?: string): Promise<SqsQueue[]> {
  const url = profile
    ? `${API_BASE}/sqs/queues?profile=${encodeURIComponent(profile)}`
    : `${API_BASE}/sqs/queues`;
  const res = await fetch(url);
  if (!res.ok) throw new Error(`Failed to fetch queues (API ${res.status})`);
  return res.json();
}

export async function sendMessage(req: SqsSendRequest): Promise<SqsSendResponse> {
  return post<SqsSendResponse>('/sqs/send', req);
}

export async function purgeQueue(
  queueUrl: string,
  confirmation: string,
  profile?: string | null,
): Promise<{ purged: string }> {
  return post<{ purged: string }>('/sqs/purge', { queueUrl, confirmation, profile });
}
